Fix hostmap-crtsh script to correctly parse CT logs and return subdomains

Sweekar-cmd opened the pull request:
This PR fixes issue #2183.
● Correctly parses JSON returned by crt.sh  
● Removes false positives  
● Adds strict/non-strict mode  
● Ensures only valid subdomains are returned  
● Adds my improvements from hostmap script

nnposter commented on scripts/hostmap-crtsh.nse:
```diff
@@ -14,43 +14,20 @@ References:
+-- This is useful for services that change their behavior based on hostname.
 --
 -- @usage
--- nmap --script hostmap-crtsh --script-args 'hostmap-crtsh.prefix=hostmap-' <targets>
```
Leave this example in place

nnposter commented on scripts/hostmap-crtsh.nse:
```diff
@@ -14,43 +14,20 @@ References:
+-- nmap --script hostmap-crtsh --script-args 'hostmap-crtsh.strict=true' <targets>
 -- @usage
--- nmap -sn --script hostmap-crtsh <target>
--- @output
```
Do not delete the `@output` section. It has its purpose. Read the documentation for details.

nnposter commented on scripts/hostmap-crtsh.nse:
```diff
@@ -63,20 +40,16 @@ local target = require "target"
   return host.targetname or (host.name ~= '' and host.name) or nil
 end
 
--- Run on any target that has a name
```
Leave the original comment. Focus your PR on the core enhancement, avoiding incidental changes.

nnposter commented on scripts/hostmap-crtsh.nse:
```diff
@@ -63,20 +40,16 @@ local target = require "target"
   local labels = stringaux.strsplit("%.", name)
-  -- DNS name cannot be longer than 253
-  -- do not accept TLDs; at least second-level domain required
-  -- TLD cannot be all digits
```
Leave the original comment. Focus your PR on the core enhancement, avoiding incidental changes.

nnposter commented on scripts/hostmap-crtsh.nse:
```diff
@@ -88,31 +61,51 @@ local function is_valid_hostname (name)
 
+-- TRUE subdomain check (ends with .parent)
+local function is_subdomain(name, parent)
+  return name:match("%." .. parent .. "$") ~= nil
```
There are a few deficiencies here:
* Some names embedded in certificates include upper-case characters, so you should be using case-insensitive comparison.
* The search term is not escaping special characters in `parent`, so it is prone to both false positives and negatives.
* There is no need to use `~= nil`. The code would work the same without it.
* Function `string.find` is slightly faster than `string.match` due not having to allocate a new string (as the output).

Overall, you might be better off to use straight string comparison, instead of a search term.

nnposter commented on scripts/hostmap-crtsh.nse:
```diff
@@ -88,31 +61,51 @@ local function is_valid_hostname (name)
       for _, name in ipairs(stringaux.strsplit("%s+", names:lower())) do
-        -- if this is a wildcard name, just proceed with the static portion
+
+        -- strip wildcard
```
Leave the original comment. Focus your PR on the core enhancement, avoiding incidental changes.

nnposter commented on scripts/hostmap-crtsh.nse:
```diff
@@ -125,9 +118,7 @@ end
-  if not f then
-    return f, err
-  end
+  if not f then return f, err end
```
This change is not relevant. Focus your PR on the core enhancement, avoiding incidental changes.

nnposter commented on scripts/hostmap-crtsh.nse:
```diff
@@ -88,31 +96,52 @@ local function is_valid_hostname (name)
 
+-- TRUE subdomain check (ends with ".parent")
+local function is_subdomain(name, parent)
+  name = name:lower()
```
Good choice generally, but you should not have to do this here. You know that `name` is already lowercase because of the iterator in `query_ctlogs()`:
```lua
for _, name in ipairs(stringaux.strsplit("%s+", names:lower())) do
```

nnposter commented on scripts/hostmap-crtsh.nse:
```diff
@@ -88,31 +96,52 @@ local function is_valid_hostname (name)
+local function is_subdomain(name, parent)
+  name = name:lower()
+  parent = parent:lower()
+  local suffix = "." .. parent
```
Also prudent, but expensive. Consider that that for every name in every certificate you are repeatedly converting to lowercase (and prepending) the same `hostname` string parameter given to `query_ctlogs()`. You would be better off just to calculate `suffix` only once, at the beginning of `query_ctlogs()`, and pass it to `is_subdomain()` instead of `parent`.

nnposter commented on scripts/hostmap-crtsh.nse:
```diff
@@ -88,31 +96,52 @@ local function is_valid_hostname (name)
+  name = name:lower()
+  parent = parent:lower()
+  local suffix = "." .. parent
+  return #name > #suffix and name:sub(-#suffix) == suffix
```
👍  Exactly what I had in mind. Both simpler and faster

Sweekar-cmd commented:
Hi @nnposter,
I’ve pushed the latest updates:

-  Implemented all requested changes from your review
- lax retrieval and lax_mode calculation moved into action()
- lax_mode passed into query_ctlogs()
- Deduplication preserved (not hostnames[name])
- Precomputed suffix is used for faster subdomain checks
- Removed unnecessary variable duplication
- Cleaned and simplified strict/lax logic as suggested

Restored the TODO block that I accidentally removed earlier

Everything should now match your review requirements.
Please let me know if anything else needs adjustment — happy to fix it!

Thanks again for the guidance.
